### PR TITLE
Quick links refactoring

### PR DESCRIPTION
**Pull request by alintreznai**
This MR refactors the quick links component

### CODE REVIEW

File: rdrf/rdrf/admin.py [hunk 1]
```diff
@@ -490,7 +490,7 @@ class CDEFileAdmin(admin.ModelAdmin):
     ADMIN_COMPONENTS = ADMIN_COMPONENTS + DESIGN_MODE_ADMIN_COMPONENTS
 
 if settings.SYSTEM_ROLE is SystemRoles.NORMAL:
-    ADMIN_COMPONENTS = ADMIN_COMPONENTS + NORMAL_MODE_ADMIN_COMPONENTS
+    ADMIN_COMPONENTS = ADMIN_COMPONENTS + NORMAL_MODE_ADMIN_COMPONENTS + PROMS_ADMIN_COMPONENTS
```
Comment (sztamas) [inline rdrf/rdrf/admin.py]: Why is this needed? We would never have PROM links in NORMAL mode.

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 2]
```diff
@@ -18,297 +18,257 @@ class QuickLink(object):
...
-            States.text: States
-        }
-        WORKING_GROUPS = {
-            WorkingGroups.text: WorkingGroups,
-        }
-        # When enabled, doctors links
-        ENABLED_DOCTORS = {
-            Doctors.text: Doctors,
-        }
+def is_clinical_or_dev():
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: Wouldn't these `is_SOME_ROLE method` would be better placed on SystemRole.

Additionally, it would make sense to have some `is_CIC` method for all the CIC roles and then we would have `is_proms` (as declared below) and `is_clinical_or_dev` would be CIC but not proms.

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 3]
```diff
@@ -18,297 +18,257 @@ class QuickLink(object):
...
-    def __setattr__(self, name):
-        return setattr(self._instance, name)
 
+class Links:
+    """
+    All links that can appear in menus are defined.
+    Links are also grouped into related functional areas to make for easier assignment to menus
+    """
+    # related links are grouped or convenience
+    AUDITING = dict([
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: Could be a tad nicer if the `dict([make_entry(QL1), ..., make_entry(QLN)])` construct would be:
`links(QL1, ... , QLN)`. etc.

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 4]
```diff
@@ -18,297 +18,257 @@ class QuickLink(object):
...
+        make_entry(LinkDefs.LoginLog),
+        make_entry(LinkDefs.FailedLoginLog),
+        make_entry(LinkDefs.LoginAttempts)
+    ]) if has_regular_roles() else {}
+
+    DATA_ENTRY = dict([
+        make_entry(LinkDefs.PatientsListing),
+    ]) if has_regular_roles() else {}
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: Aren't we repeating the `if has_regular_roles()` etc. again in `QuickLinks._build_menu`?

It looks like we could omit all the ifs here if we are going to have them in `QuickLinks._build_menu`.

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 5]
```diff
@@ -18,297 +18,257 @@ class QuickLink(object):
...
+        make_entry(LinkDefs.Users),
+    ])
+
+    PROMS = dict([
+        make_entry(LinkDefs.Surveys),
+        make_entry(LinkDefs.SurveyAssignments),
+        make_entry(LinkDefs.SurveyRequest)
+    ]) if supports_proms_role() else {}
+
+    REGISTRY_DESIGN = dict([
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: Isn't this repeated above?

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 6]
```diff
@@ -402,27 +360,28 @@ def _build_menu(self):
...
+            RDRF_GROUPS.CLINICAL: self.menu_config.clinical,
+            RDRF_GROUPS.GENETIC_STAFF: self.menu_config.genetic_staff,
+            RDRF_GROUPS.GENETIC_CURATOR: self.menu_config.genetic_curator,
+            RDRF_GROUPS.WORKING_GROUP_STAFF: self.menu_config.working_group_staff,
+            RDRF_GROUPS.WORKING_GROUP_CURATOR: self.menu_config.working_group_curator,
+            RDRF_GROUPS.SUPER_USER: self.menu_config.super_user,
         }
         return switcher.get(group.lower(), [])
 
     def __init__(self, registries):
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: `__init__` method should come before the other methods.

**Review by sztamas — COMMENTED** [review]
Added some inline comments.

Additionally, it might worthwhile doing an experiment with using polymorphism instead of all those ifs.
That might be more readable as you would go to the right class to see a list of links (settings, menu and admin_page) that would be presented.
We would have different classes based on SystemRole, registries in the system, and user group.

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 7]
```diff
@@ -402,27 +360,28 @@ def _build_menu(self):
...
-            if settings.DESIGN_MODE:
+            if is_design_mode():
                 design_menus.update(proms_menus)
-                MenuConfig().all = design_menus
+                self.menu_config.all = design_menus
             else:
-                MenuConfig().all = proms_menus
+                self.menu_config.all = proms_menus
 
     def _group_links(self, group):
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: Is this used anywhere? Even if it is, why isn't this a method on `MenuConfig` instead?

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 8]
```diff
@@ -350,32 +308,32 @@ def _build_menu(self):
...
             }
 
-            MenuConfig().clinical = {
+            self.menu_config.clinical = {
                 **Links.DATA_ENTRY,
                 **Links.QUESTIONNAIRE,
                 **Links.VERIFICATION,
             }
 
             # Super user has combined menu of all other users
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: Looks like a strange idea to me.
ie. that adding up menus of users with different roles would necessarily add up to all the menus in the system...

File: rdrf/rdrf/forms/navigation/quick_links.py [hunk 9]
```diff
@@ -385,14 +343,14 @@ def _build_menu(self):
             }
 
             # menu with everything, used for the admin page
-            if not settings.DESIGN_MODE:
-                MenuConfig().all = normal_menus
+            if not is_design_mode():
```
Comment (sztamas) [inline rdrf/rdrf/forms/navigation/quick_links.py]: `if/else` with a `not` condition. Any reason not to switch around and get rid of the `not`?

Or perhaps even better with less branching:

```python
self.menu.config.all = normal_menus
if is_design_mode():
    self.menu.config.all.update(design_menus)
```

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 7 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Yes. It's used in the menu_links method. Moved it to MenuConfig.

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 3 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Updated.

File: rdrf/rdrf/admin.py (hunk 1 above)
Comment (alintreznai) [reply rdrf/rdrf/admin.py]: It's a leftover before I changed url resolving using reverse_lazy. Removed the change.

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 5 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Indeed. Removed.

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 6 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Updated.

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 9 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Updated.

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 2 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Updated.

File: rdrf/rdrf/forms/navigation/quick_links.py (hunk 4 above)
Comment (alintreznai) [reply rdrf/rdrf/forms/navigation/quick_links.py]: Updated.

**Comment by alintreznai** [comment]
Updated the code to use polymorphism.
